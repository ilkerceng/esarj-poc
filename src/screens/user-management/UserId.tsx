import { useMemo } from 'react';
import { ID } from '../../lib/types';

const leadingZerosPattern = '[^0]+';

export const splitWithLeadingPattern = (
  value: string,
  pattern = leadingZerosPattern,
): { leadingStr: string; otherPart: string } | undefined => {
  if (value) {
    const idStr = value.toString();
    const firstNonZeroNumIndex = value.toString().search(pattern);
    const leadingStr = idStr.substring(0, firstNonZeroNumIndex);
    const otherPart = idStr.substring(firstNonZeroNumIndex);

    return { leadingStr, otherPart };
  }
};

export const UserID = ({ id }: { id: ID }) => {
  const result = useMemo(() => {
    return splitWithLeadingPattern(id?.toString());
  }, [id]);

  return result ? (
    <>
      <span className="opacity-20">{result.leadingStr}</span>
      <span>{result.otherPart}</span>
    </>
  ) : null;
};
