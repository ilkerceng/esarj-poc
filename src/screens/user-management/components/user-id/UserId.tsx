import { HTMLProps, useMemo } from 'react';
import { ID } from '../../../../lib/types';
import { splitWithLeadingPattern } from '../../../../lib/utils';

type UserIDProps = {
  id: ID;
  leadingZerosContainerProps?: HTMLProps<HTMLSpanElement>;
  numericIdContainerProps?: HTMLProps<HTMLSpanElement>;
};

/**
 * Renders user id with less opacity leading zeros.
 */
export const UserID = ({
  id,
  leadingZerosContainerProps,
  numericIdContainerProps,
}: UserIDProps) => {
  const result = useMemo(() => {
    return splitWithLeadingPattern(id?.toString());
  }, [id]);

  return result ? (
    <>
      <span className="opacity-20" {...leadingZerosContainerProps}>
        {result.leadingStr}
      </span>
      <span {...numericIdContainerProps} >{result.otherPart}</span>
    </>
  ) : null;
};
