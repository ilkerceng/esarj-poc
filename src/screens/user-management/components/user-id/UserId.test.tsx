import { render } from '@testing-library/react';
import { UserID } from './UserId';
import renderer from 'react-test-renderer';

test('renders correctly', () => {
  const tree = renderer.create(<UserID id="00123" />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('set leading zeros opacity correctly', () => {
  const { container } = render(<UserID id="00123" />);
  expect(container.firstChild).toHaveClass('opacity-20');
});

test('renders id having leading zeros correctly', () => {
  const { getByText } = render(<UserID id="00123" />);

  const leadingZeros = getByText(/00/i);
  expect(leadingZeros).toBeInTheDocument();

  const otherParts = getByText(/123/i);
  expect(otherParts).toBeInTheDocument();
});

test('renders id does not have any leading zeros correctly', () => {
  const { getByText, queryByTestId } = render(<UserID id="12345" />);

  const leadingZeros = queryByTestId('leading-zeros');
  expect(leadingZeros).not.toBeInTheDocument();

  const otherParts = getByText(/12345/i);
  expect(otherParts).toBeInTheDocument();
});
