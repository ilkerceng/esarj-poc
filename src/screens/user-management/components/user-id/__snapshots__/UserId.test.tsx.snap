// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
[
  <span
    className="opacity-20"
    data-testid="leading-zeros"
  >
    00
  </span>,
  <span>
    123
  </span>,
]
`;
